test_util/src/lib.rs

```diff
@@ -108,44 +123,242 @@ pub fn generate_aead_key() -> LessSafeKey {
+    pub async fn wait_for_sleeping_tasks(&self, num_tasks: usize) {
+        let mut receiver = {
+            let waiting_tasks_guard = self.inner.waiting_tasks.lock().await;
+            // Early-out check
+            if waiting_tasks_guard.0.len() >= num_tasks {
+                return;
+            }
+            waiting_tasks_guard.1.subscribe()
+        };
+
+        loop {
+            // Wait to be notified that another task is sleeping.
+            receiver.recv().await.unwrap();
+            let waiting_tasks_guard = self.inner.waiting_tasks.lock().await;
+            if waiting_tasks_guard.0.len() >= num_tasks {
+                break;
+            }
+        }
     }
 }
 
+#[async_trait]
 impl Clock for MockClock {
+    type Interval = MockInterval;
+
     fn now(&self) -> Time {
-        let current_time = self.current_time.lock().unwrap();
-        *current_time
+        let current_times = self.inner.current_times_std_mutex.lock().unwrap();
+        current_times.0
+    }
+
+    fn now_monotonic(&self) -> Instant {
+        let current_times = self.inner.current_times_std_mutex.lock().unwrap();
+        current_times.1
+    }
+
+    async fn timeout<O, F>(&self, duration: StdDuration, future: F) -> Result<O, Elapsed>
+    where
+        F: Future<Output = O> + Send,
+    {
+        let deadline = self.now_monotonic() + duration;
+        MockTimeoutCancellable::new(future, || self.now_monotonic() >= deadline).await
+    }
+
+    fn interval_at(&self, start: Instant, period: StdDuration) -> Self::Interval {
+        MockInterval {
+            next_tick: start,
+            period,
+            clock: self.clone(),
+        }
+    }
+
+    async fn sleep(&self, duration: StdDuration) {
+        let deadline = self.now_monotonic() + duration;
+        self.sleep_until(deadline).await;
     }
 }
 
 impl Default for MockClock {
     fn default() -> Self {
-        Self {
-            // Sunday, September 9, 2001 1:46:40 AM UTC
-            current_time: Arc::new(Mutex::new(Time::from_seconds_since_epoch(1000000000))),
+        // Sunday, September 9, 2001 1:46:40 AM UTC
+        Self::new(Time::from_seconds_since_epoch(1000000000))
+    }
+}
+
+/// An interval used in conjunction with the [`MockClock`].
+#[derive(Debug)]
+pub struct MockInterval {
+    next_tick: Instant,
+    period: StdDuration,
+    clock: MockClock,
+}
+
+#[async_trait]
+impl ClockInterval for MockInterval {
+    async fn tick(&mut self) {
+        self.clock.sleep_until(self.next_tick).await;
+        self.next_tick += self.period;
+    }
+
+    fn set_missed_tick_behavior(&mut self, _behavior: MissedTickBehavior) {
+        // Ignore this setting, because time doesn't advance smoothly with a MockClock.
+    }
+}
+
+/// This future combinator is used by `MockClock::timeout` to check the timeout at every yield
+/// point, and cancel if the timeout has elapsed.
+#[pin_project]
+struct MockTimeoutCancellable<FU, FN, O>
```

Comment:
I think we don't need to roll our own future here: we could instead have `MockClock::timeout` return an async block which `select!`s on the inner future & a call to `sleep_until` (with the branches being translated to the appropriate result/error types). Given the complexity of correctly implementing `poll`, we should avoid doing so unless we need to IMO.